source/common/common/packed_struct.h

```diff
@@ -0,0 +1,139 @@
+    static_assert(std::is_enum_v<ElementName>);
+    static_assert(max_size > 0);
+
+    RELEASE_ASSERT(a_capacity < std::numeric_limits<uint8_t>::max(),
+                   "capacity should fit in uint8_t.");
+    indices_.fill(max_size);
+    indices_[max_size] = a_capacity;
+  }
+
+  T& get(ElementName element_name) {
+    return (*this)[static_cast<std::underlying_type_t<ElementName>>(element_name)];
+  }
+  const T& get(ElementName element_name) const {
+    return (*this)[static_cast<std::underlying_type_t<ElementName>>(element_name)];
+  }
+  bool has(ElementName element_name) const {
+    return has(static_cast<std::underlying_type_t<ElementName>>(element_name));
+  }
+  void set(ElementName element_name, T t) {
+    return assign(static_cast<std::underlying_type_t<ElementName>>(element_name), t);
+  }
+
+  // Number of non-empty elements in the struct. Note that this can be less than
+  // capacity.
+  size_t size() const {
+    return std::accumulate(indices_.begin(), indices_.end() - 1, 0,
+                           [](uint8_t a, uint8_t b) { return b < max_size ? ++a : a; });
+  }
+
+  size_t capacity() const { return indices_[max_size]; }
+
+  // Disable copying
+  PackedStruct(const PackedStruct&) = delete;
+
+  // Move constructor and assignment operator.
+  PackedStruct(PackedStruct&& other) noexcept = default;
+  PackedStruct& operator=(PackedStruct&& other) = default;
+
+private:
+  // Accessors.
+  T& operator[](size_t idx) {
+    RELEASE_ASSERT(idx < static_cast<size_t>(max_size),
+                   absl::StrCat("idx (", idx, ") should be less than ", max_size));
+    RELEASE_ASSERT(indices_[idx] < max_size,
+                   absl::StrCat("Element corresponding to index ", idx, " is not assigned"));
+    return (data_.get())[indices_[idx]];
+  }
+  const T& operator[](size_t idx) const {
+    RELEASE_ASSERT(idx < static_cast<size_t>(max_size),
+                   absl::StrCat("idx (", idx, ") should be less than ", max_size));
+    RELEASE_ASSERT(indices_[idx] < max_size,
+                   absl::StrCat("Element corresponding to index ", idx, " is not assigned"));
+    return (data_.get())[indices_[idx]];
+  }
+
+  bool has(size_t idx) const {
+    RELEASE_ASSERT(idx < static_cast<size_t>(max_size),
+                   absl::StrCat("idx (", idx, ") should be less than ", max_size));
+    return indices_[idx] < max_size;
+  }
+
+  void assign(size_t element_idx, T t) {
+    RELEASE_ASSERT(element_idx < static_cast<size_t>(max_size),
+                   absl::StrCat("element_idx (", element_idx, ") should be less than ", max_size));
+    auto const current_size = size();
+    // If we're at capacity and we don't have a slot for element_idx, increase capacity by 1.
+    if (!has(element_idx) && current_size == capacity()) {
+      std::unique_ptr<T[]> tmp(new T[++indices_[max_size]]);
+      std::move(data_.get(), std::next(data_.get(), current_size), tmp.get());
+      swap(data_, tmp);
```

Comment:
would data_ = std::move(tmp) work?

Reply:
Yeah I was just suggesting to swap the swap for a move, as we don't care about the residual value in tmp.